README.md
# fuse

This is an old version of libfuse initially released [here](https://github.com/LineageOS/android_external_fuse/tree/cm-14.1), which was patched with android support. It has been further patched for NDK cross compilation and Termux native compilation. The device kernel must support the fuse filesystem. Check "/proc/filesystems" for "fuse" entry. Linux kernels 2.6.14 or later
should contain FUSE support.

## Warning: unresolved security issue

Be aware that FUSE has an unresolved security bug
([bug #15](https://github.com/libfuse/libfuse/issues/15)): the
permission check for accessing a cached directory is only done once
when the directory entry is first loaded into the cache. Subsequent
accesses will re-use the results of the first check, even if the
directory permissions have since changed, and even if the subsequent
access is made by a different user.

This bug needs to be fixed in the Linux kernel and has been known
since 2006 but unfortunately no fix has been applied yet. If you
depend on correct permission handling for FUSE file systems, the only
workaround is to completely disable caching of directory
entries. Alternatively, the severity of the bug can be somewhat
reduced by not using the `allow_other` mount option.


## About

FUSE (Filesystem in Userspace) is an interface for userspace programs
to export a filesystem to the Linux kernel. The FUSE project consists
of two components: the *fuse* kernel module (maintained in the regular
kernel repositories) and the *libfuse* userspace library (maintained
in this repository). libfuse provides the reference implementation
for communicating with the FUSE kernel module.

A FUSE file system is typically implemented as a standalone
application that links with libfuse. libfuse provides functions to
mount the file system, unmount it, read requests from the kernel, and
send responses back. libfuse offers two APIs: a "high-level",
synchronous API, and a "low-level" asynchronous API. In both cases,
incoming requests from the kernel are passed to the main program using
callbacks. When using the high-level API, the callbacks may work with
file names and paths instead of inodes, and processing of a request
finishes when the callback function returns. When using the low-level
API, the callbacks must work with inodes and responses must be sent
explicitly using a separate set of API functions.


## Compile and Install Instructions for Linux Distros

It is best to go [here](https://github.com/libfuse/libfuse) for latest libfuse versions for pc distros and use those.
```
#install dependencies
sudo apt install autoconf automake libtool

cd fuse

#only if building from git repository to generate configure script
./makeconf.sh

#build
./configure
make -j8

#install if needed
make install
```

You may also need to add `/usr/local/lib` to `/etc/ld.so.conf` and/or
run *ldconfig*. If you're building from the git repository (instead of
using a release tarball), you also need to run `./makeconf.sh` to
create the `configure` script.

For more details see the file `INSTALL`


## Cross Compile and Install Instructions for Android Using NDK

- Download [android_ndk_cross_compile_build_automator](https://github.com/agnostic-apollo/Android-NDK-Cross-Compile-Build-Automator) and read/follow its usage guide.

- Copy `fuse` directory to `android_ndk_cross_compile_build_automator/packages` directory.

- Copy `fuse/post_build_scripts/fusermount_extractor.sh` file to `android_ndk_cross_compile_build_automator/$POST_BUILD_SCRIPTS_DIR/` directory.

- Add/Set `fuse` to `$PROJECTS_TO_BUILD` in `android_ndk_cross_compile_build_automator.sh`.

- Add/Set `fusermount_extractor.sh` to `$POST_BUILD_SCRIPTS_TO_RUN` in `android_ndk_cross_compile_build_automator.sh`. You can skip this optionally if you do not want to create a zip of the fusermount binaries of all the archs that are built.

- Add/Set `armeabi armeabi-v7a arm64-v8a x86 x86-64` to `$ARCHS_SRC_TO_BUILD` in `android_ndk_cross_compile_build_automator.sh` or whatever ARCHS_SRC you want to build for. API_LEVEL in ARCH_SRC files must be 21 or higher otherwise compilation will fail. NDK must also be higher than r15. Check `fuse/android_ndk_cross_compile_build.sh` for more details.

```
#install dependencies
sudo apt install autoconf automake libtool

cd android_ndk_cross_compile_build_automator

#build
bash ./android_ndk_cross_compile_build_automator

```

- `android_ndk_cross_compile_build_automator` will call `fuse/android_ndk_cross_compile_build.sh script` to build and install each ARCHS_SRC. fuse for each ARCHS_SRC will be installed at `android_ndk_cross_compile_build_automator/$INSTALL_DIR/fuse/$ARCHS_SRC`.

- `android_ndk_cross_compile_build_automator/$POST_BUILD_SCRIPTS_DIR/fusermount_extractor.sh` will extract `fusermount` binaries from `android_ndk_cross_compile_build_automator/$INSTALL_DIR/fuse/$ARCHS_SRC` and zip them at `android_ndk_cross_compile_build_automator/$OUT_DIR/fuse/fusermount<build-info>.zip`.


Fully static compile is not possible currently for fusermount binary using NDK since bionic libc does not support static linking (fully or at all), since binaries built for android are supposed to dynamically link with android system libraries of each phone and android version at runtime. Fusermount compiled with NDK will dynamically link with android system libc.so and/or libdl.so.
However the fusermount compiled with above flags is not linked with any termux libraries and should not need the export of termux lib path with LD_LIBRARY_PATH for execution. Fully static compile will probably be possible with mucl since its libc implementation has static support. A focus on only the fusermount binary is because it is required by [rclone](https://rclone.org) [mount](https://rclone.org/commands/rclone_mount) for mounting cloud drives like google drive in android with root of course. That was the motivation behind this all. 


## Native Compile and Install Instructions for Android on Android Using Termux

```
#install dependencies
pkg install build-essential git silversearcher-ag wget gettext

cd fuse

#must be run to generate configure script and fulfill dependencies in termux
./makeconf_termux.sh

#build
./configure CC=clang CXX=clang++ --enable-example=no --disable-mtab --enable-android=yes --enable-static=yes LDFLAGS="-static-libgcc"
make

#install fusermount by copying it to termux bin path and set correct ownership and permissions
cp ./util/fusermount /data/data/com.termux/files/usr/bin
export scripts_path="/data/data/com.termux/files/usr/bin"; export termux_uid="$(id -u)"; export termux_gid="$(id -g)"; su -c chown $termux_uid:$termux_gid "$scripts_path/fusermount" && chmod 700 "$scripts_path/fusermount";

```

Fully static compile is not possible currently for fusermount binary using termux since termux clang does not support static compilations and even if that were possible, fusermount would most likely still link dynamically with android system libc.so and/or libdl.so.
However the fusermount compiled with above flags is not linked with any termux libraries and should not need the export of termux lib path with LD_LIBRARY_PATH for execution.  

You can check out [Termux Build Packages Wiki](https://wiki.termux.com/wiki/Building_packages) for more helpful info.


## Security implications

If you run `make install`, the *fusermount* program is installed
set-user-id to root.  This is done to allow normal users to mount
their own filesystem implementations.

There must however be some limitations, in order to prevent Bad User from
doing nasty things.  Currently those limitations are:

  - The user can only mount on a mountpoint, for which it has write
    permission

  - The mountpoint is not a sticky directory which isn't owned by the
    user (like /tmp usually is)

  - No other user (including root) can access the contents of the
    mounted filesystem (though this can be relaxed by allowing the use
    of the `allow_other` and `allow_root` mount options in `fuse.conf`)


## Building your own filesystem

FUSE comes with several example file systems in the `examples`
directory. For example, the *fusexmp* example mirrors the contents of
the root directory under the mountpoint. Start from there and adapt
the code!

The documentation of the API functions and necessary callbacks is
mostly contained in the files `include/fuse.h` (for the high-level
API) and `include/fuse_lowlevel.h` (for the low-level API). An
autogenerated html version of the API is available in the `doc/html`
directory and at http://libfuse.github.io/doxygen.


## Getting Help

If you need help, please ask on the <fuse-devel@lists.sourceforge.net>
mailing list (subscribe at
https://lists.sourceforge.net/lists/listinfo/fuse-devel).

Please report any bugs on the GitHub issue tracker at
https://github.com/libfuse/main/issues.



ChangeLog
FUSE 2.9.7 (2019-08-1)
=======================
* Patched by: agnostic-apollo
* Added makeconf_termux.sh for termux compilation
* Added `--enable-android` `AC_ARG_ENABLE` in `configure.ac` for setting `__ANDROID__` macro since that was not already being set and was needed so that mtab is never used among other things. `__ANDROID__` macro also ensures that mount and umount `--fake` and `--no-canonicalize` command options are never used by making sure that `mtab_needs_update` function always returns `0` when called in `fuse_mnt_add_mount` and `fuse_mnt_remove_mount` functions in `lib/mount_util.c`. This is done because android system mount and umount binaries do not support those command options.
* Added `m4_ifdef` when `AM_ICONV` is called in `configure.ac`, since this is not defined when compiling in termux since termux `gettext` package does not provide `iconv.m4`.
* Also added condition in `configure.ac` that if `--enable-android` is passed, then do not check for mount and umount `--fake` and `--no-canonicalize` command options support.

FUSE 2.9.7 (2016-06-20)
=======================

* Added SELinux support.
* Fixed race-condition when session is terminated right after starting
  a FUSE file system.

FUSE 2.9.6 (2016-04-23)
=======================

* Tarball now includes documentation.
* Shared-object version has now been bumped correctly.

FUSE 2.9.5 (2016-01-14)
=======================

* New maintainer: Nikolaus Rath <Nikolaus@rath.org>. Many thanks to
  Miklos Szeredi <miklos@szeredi.hu> for bringing FUSE to where it is
  now!

* fix warning in mount.c:receive_fd().  Reported by Albert Berger

* fix possible memory leak.  Reported by Jose R. Guzman

FUSE 2.9.4 (2015-05-22)
=======================

* fix exec environment for mount and umount.  Found by Tavis Ormandy
  (CVE-2015-3202).

* fix fuse_remove_signal_handlers() to properly restore the default
  signal handler.  Reported by: Chris Johnson

* highlevel API: fix directory file handle passed to ioctl() method.
  Reported by Eric Biggers

* libfuse: document deadlock avoidance for fuse_notify_inval_entry()
  and fuse_notify_delete()

* fusermount, libfuse: send value as unsigned in "user_id=" and
  "group_id=" options.  Uids/gids larger than 2147483647 would result
  in EINVAL when mounting the filesystem.  This also needs a fix in
  the kernel.

* Initilaize stat buffer passed to ->getattr() and ->fgetattr() to
  zero in all cases.  Reported by Daniel Iwan

* libfuse: Add missing includes.  This allows compiling fuse with
  musl.  Patch by Daniel Thau


Older Versions (before 2013-01-01)
==================================


2013-06-20  Miklos Szeredi <miklos@szeredi.hu>

	* libfuse: fix multiple close of device fd.  Reported by Dan
	Greenfield

2013-03-19  Miklos Szeredi <miklos@szeredi.hu>

	* libfuse: fix thread cancel race.  Exiting a worker my race with
	cancelling that same worker.  This caused a segmenation
	fault. Reported and tested by Anatol Pomozov

2013-02-04  Miklos Szeredi <miklos@szeredi.hu>

	* libfuse: fix crash in unlock_path().  Patch by Ratna Manoj

	* libfuse: fix the 'remember' option.  The lru list was not
	initialized for the "/" path.  This resulted in remove_node_lru()
	crashing on LOOKUP-DOTDOT.  Patch by Madan Valluri

	* libfuse: configure: detect new util-linux

	* libfuse: Use AC_CONFIG_HEADERS instead of AM_CONFIG_HEADER.
	Patch by Anatol Pomozov

	* libfuse: rename ./configure.in to ./configure.ac.  Patch by
	Anatol Pomozov

2012-10-01  Miklos Szeredi <miklos@szeredi.hu>

	* Released 2.9.2

2012-10-01  Miklos Szeredi <miklos@szeredi.hu>

	* Fix deadlock in libfuse.  Running "svn update" on a fuse
	filesystem could deadlock because of a bug in the way the paths
	are locked.  Reported by Kazuaki Anami

2012-08-23  Miklos Szeredi <miklos@szeredi.hu>

	* Fix missing config.h in buffer.c.  Reported by Matthew Gabeler-Lee

2012-08-14  Miklos Szeredi <miklos@szeredi.hu>

	* Not unhashing the name in forget (commit on 2011-12-09) broke
	the forget logic in a subtle way, resulting in "fuse internal
	error: node NNN not found" and causing the filesystem daemon to
	abort.  Fix by incrementing the node refcount if nlookup goes from
	zero to one.  Reported by Kyle Lippincott

2012-08-13  Miklos Szeredi <miklos@szeredi.hu>

	* Fix linking against GNU libiconv.  Patch by Natanael Copa

2012-07-19  Miklos Szeredi <miklos@szeredi.hu>

	* Released 2.9.1

2012-07-19  Miklos Szeredi <miklos@szeredi.hu>

	* Fix crash caused by freeing a stack address.  Reported by Itay
	Perl

2012-07-04  Miklos Szeredi <miklos@szeredi.hu>

	* Fix install of mount.fuse from out-of-tree build.  Patch by
	Olivier Blin

	* Fix build with automake >= 1.12.1.  Patch by Olivier Blin

2012-04-24  Miklos Szeredi <miklos@szeredi.hu>

	* Add fallocate operation.  Only works on linux kernels 3.5 or
	later.  Patch by Anatol Pomozov

2012-05-16  Miklos Szeredi <miklos@szeredi.hu>

	* Linking to a library that uses threads requires the application
	to be linked with -pthreads otherwise some pthread functions will
	be linked to stubs in glibc.  So move -pthread from Libs.private
	to Libs in fuse.pc.  Reported by Werner Fink

	* Fix the compile command in the examples. Reported by Luciano
	Dalle Ore

2012-04-20  Miklos Szeredi <miklos@szeredi.hu>

	* Released 2.9.0

2012-04-20  Miklos Szeredi <miklos@szeredi.hu>

	* Add missing fuse_fs_flock to fuse_versionscript

2012-04-10  Miklos Szeredi <miklos@szeredi.hu>

	* Check protocol version before sending notifications and return
	-ENOSYS if a particular notification is not supported.

	* Add 'flag_utime_omit_ok' flag to fuse_operations.  If the
	filesystem sets this flag then ->utimens() will receive UTIME_OMIT
	and UTIME_NOW values as specified in utimensat(2).

2012-01-27  Miklos Szeredi <miklos@szeredi.hu>

	* Interpret octal escape codes in options.  Requested by Jan
	Engelhardt

2012-01-26  Miklos Szeredi <miklos@szeredi.hu>

	* Add man pages for fusermount, mount.fuse and ulockmgr_server.

android_ndk_cross_compile_build.sh
#!/bin/bash
#title:          android_ndk_cross_compile_build.sh
#description:    NDK cross compile build script for fuse using an autoconf build system	
#author:         agnostic-apollo
#usage:          To be used with `android_ndk_cross_compile_build_automator.sh`.
#date:           1-Aug-2019
#versions:       1.0
#license:        MIT License


#setfsuid and setfsgid used in util/fusermount.c are GNU Glibc functions which were added to Android bionic in Android 5.0 (API level 21)
#Building for API level less than 21 will fail with undefined fuction errors
#the stat struct containing the members st_atim and st_mtim which store nanosecond timestamps is used in lib/fuse.c,
#but ndk older than r15 did not have these implemented and so so compilation will fail on older versions (might be possible with r14 by including unified headers)
#Since fusermount can only be currently compiled for the minimum API 21,
#running it on devices having API level less than 21 could result in problems and would need testing
#Otherwise the missing functions might need to be ported if at all possible
#Moreover building with NDK still results in dynamic binding for libc.so for 64 bit binaries and libc.so and libdl.so for 32 bit binaries
#If that causes problems in some devices, then mucl could be looked into for static compilation instead of NDK


#The following variables are exported by the android_ndk_cross_compile_build_automator.sh script before calling this script
#ARCH_SRC: The name of the ARCH_SRC this project is being built for
#API_LEVEL: The API_LEVEL this project is being built for, toolchain is already generated for this API level
#ABI_NAME: The abi of the arch this project is being built for
#TARGET_HOST: The host this project should be built for. It should contain any value in "arm-linux-androideabi \
#armv7a-linux-androideabi aarch64-linux-android i686-linux-android x86_64-linux-android mips64el-linux-android \
#mips64el-linux-android"

#PROJECT: The name of the project directory being built
#BUILD_DIR: The absolute path to the parent directory of the project
#INSTALL_DIR: The absolute path to the directory of that can be used as prefix for install commands
#TOOLCHAIN_DIR: The absolute path to the toolchain directory
#SYSROOT: The absolute path to the toolchain SYSROOT directory
#TOOLCHAIN_BIN_DIR: The absolute path to the toolchain bin directory
#The "CPP CC CXX LD AR AS NM RANLIB STRIP" variables are also exported depending on the toolchain

#C_COMPILER: The c compiler of the toolchain. It should contain any value in "gcc clang"
#CXX_COMPILER: The c compiler of the toolchain. It should contain any value in "g++ clang++"
#CFLAGS: The arch specific c compiler flags defined in ARCH_SRC file and any CFLAGS exported \
#before running android_ndk_cross_compile_build_automator.sh script
#CPPFLAGS: The arch specific c++ compiler flags defined in ARCH_SRC file and any CPPFLAGS exported \
#before running android_ndk_cross_compile_build_automator.sh script
#LDFLAGS: The arch specific linker flags defined in ARCH_SRC file and any LDFLAGS exported \
#before running android_ndk_cross_compile_build_automator.sh script


#set an ARCH_SRC that this project cannot be built for
UNSUPPORTED_ARCH_SRC="armeabi-android4.0.4-"

#fuse is compiled with a c compiler, but you will get compilation errors with gcc with ndk older than r15
#Android NDK provides gcc below NDK version 18 and clang in NDK versions greater or equal to 15
#SUPPORTED_C_COMPILER="gcc clang"
SUPPORTED_C_COMPILER="gcc clang"

#Android NDK provides g++ below NDK version 18 and clang++ in NDK versions greater or equal to 15
#SUPPORTED_C_COMPILER="g++ clang++"
SUPPORTED_CXX_COMPILER="g++ clang++"

#set ARCH_SRC for which not to add PIE/PIC flags 
ARCH_SRC_NOT_TO_ADD_PIE_PIC_FLAGS_FOR="armeabi-android4.0.4-"


#if ARCH_SRC is listed in UNSUPPORTED_ARCH_SRC, then exit with non catastrophic exit code $UNSUPPORTED_ARCH_SRC_EXIT_CODE
if [[ "$UNSUPPORTED_ARCH_SRC" =~ (^|[[:space:]])"$ARCH_SRC"($|[[:space:]]) ]]; then
	echo "$PROJECT cannot be built for $ARCH_SRC"
	exit $UNSUPPORTED_ARCH_SRC_EXIT_CODE
fi

#if C_COMPILER is not listed in SUPPORTED_C_COMPILER, then exit with non catastrophic exit code $UNSUPPORTED_C_COMPILER_EXIT_CODE
if [[ ! "$SUPPORTED_C_COMPILER" =~ (^|[[:space:]])"$C_COMPILER"($|[[:space:]]) ]]; then
	echo "$PROJECT cannot be built with $C_COMPILER"
	exit $UNSUPPORTED_C_COMPILER_EXIT_CODE
fi

#if CXX_COMPILER is not listed in SUPPORTED_CXX_COMPILER, then exit with non catastrophic exit code $UNSUPPORTED_CXX_COMPILER_EXIT_CODE
if [[ ! "$SUPPORTED_CXX_COMPILER" =~ (^|[[:space:]])"$CXX_COMPILER"($|[[:space:]]) ]]; then
	echo "$PROJECT cannot be built with $CXX_COMPILER"
	exit $UNSUPPORTED_CXX_COMPILER_EXIT_CODE
fi


#fuse uses a c compiler and needs to be compiled with static flags
#set c compiler specific flags
if [[ "$C_COMPILER" == "gcc" ]]; then
	CPPFLAGS="$CPPFLAGS"
	CFLAGS="$CFLAGS -static"
	LDFLAGS="$LDFLAGS -static -static-libgcc"
		:
elif [[ "$C_COMPILER" == "clang" ]]; then
	CFLAGS="$CFLAGS -static"
	LDFLAGS="$LDFLAGS -static -static-libgcc"
	:
fi

#fuse does not use c++ compiler, so not used, just a proof of concept
#set c++ compiler specific flags
if [[ "$CXX_COMPILER" == "g++" ]]; then
	#CPPFLAGS="$CPPFLAGS"
	#CXXFLAGS="$CFLAGS -static"
	#LDFLAGS="$LDFLAGS -static -static-libstdc++"
	:
elif [[ "$CXX_COMPILER" == "clang++" ]]; then
	#CXXFLAGS="$CFLAGS -static"
	#LDFLAGS="$LDFLAGS -static -static-libstdc++"
	:
fi


#executables or libraries built without PIE/PIC flags wont work on android-21 (Android 5.0) or higher
#executables or libraries built with PIE/PIC flags will only work on android-16 (Android 4.1) or higher
#executables or libraries to be run on android-15 (Android 4.0.4) or lower must be built without PIE/PIC flags,
#otherwise there will be runtime failure
#you can use the special ARCH_SRC file for these cases
#currently "armeabi-android4.0.4-" is for this case and can be used possibly with the same flags as "armeabi" other than API_LEVEL

#flags are passed automatically by NDK to clang
#if specific flags are to be added depending on ARCH_SRC
#if ARCH_SRC is not listed in ARCH_SRC_NOT_TO_ADD_PIE_PIC_FLAGS_FOR, then add PIE/PIC flags
if [[ ! "$ARCH_SRC_NOT_TO_ADD_PIE_PIC_FLAGS_FOR" =~ (^|[[:space:]])"$ARCH_SRC"($|[[:space:]]) ]]; then
	#CFLAGS="$CFLAGS -fPIE -fPIC"
	#LDFLAGS="$LDFLAGS -pie"
	:
fi


#cd to project root just in case
cd "$BUILD_DIR/$PROJECT"

#not needed for fuse, read ahead
#If your autoconf project does not a configure script in the source code then run autoreconf to generate it from configure.ac
#autoreconf -i
#if [ $? -ne 0 ]; then
#	echo "Failure while running autoreconf for $PROJECT for $ARCH_SRC"
#	exit 1
#fi

#fuse provides its own script makeconf.sh for running libtoolize, autoreconf and for getting other dependencies
#if your project provides its own autoconf or build script, run it
echo -e "\n\nRunning makeconf for $PROJECT for $ARCH_SRC"
./makeconf.sh
if [ $? -ne 0 ]; then
	echo "Failure while running makeconf for $PROJECT for $ARCH_SRC"
	exit 1
fi


#run configure command, you can add any additional flags
echo -e "\n\nRunning configure for $PROJECT for $ARCH_SRC"
./configure \
	--host="$TARGET_HOST" \
	--with-sysroot="$SYSROOT" \
	--enable-example=no \
	--disable-mtab \
	--enable-android=yes \
	--enable-static=yes \
	"$@"
if [ $? -ne 0 ]; then
	echo "Failure while running configure for $PROJECT for $ARCH_SRC"
	exit 1
fi


#get total cpu/cores
cpus="$(nproc)"
if [ $? -eq 0 ]; then
	make_jobs"-j$cpus"
fi


#you can optionally skip this depending on your project
#but probably might be better to have a clean build directory before building for different archs
#run make clean command
echo -e "\n\nRunning make clean for $PROJECT for $ARCH_SRC"
make $make_jobs clean
if [ $? -ne 0 ]; then
	echo "Failure while running make clean for $PROJECT for $ARCH_SRC"
	exit 1
fi

#run make clean command
echo -e "\n\nRunning make for $PROJECT for $ARCH_SRC"
make $make_jobs
if [ $? -ne 0 ]; then
	echo "Failure while running make clean for $PROJECT for $ARCH_SRC"
	exit 1
fi


project_install_dir_for_arch="$INSTALL_DIR/$PROJECT/$ARCH_SRC"

#run make install command
echo -e "\n\nRunning make install for $PROJECT for $ARCH_SRC"
make $make_jobs DESTDIR="$project_install_dir_for_arch" install
if [ $? -ne 0 ]; then
	echo "Failure while running make install for $PROJECT for $ARCH_SRC"
	exit 1
fi


echo -e "\n\nInstalled at $project_install_dir_for_arch"
echo "Complete"

